### ISSUE TITLE
Handling of ties

### ISSUE THREAD

**Issue by kno10** [rlaboiss/debian-gr-systemd-2019#5]
IMHO, ties in ballots need to be taken care of more carefully. I suggest to add these lines to your preprocessing:
```
        choices = scipy.stats.rankdata(choices)
        # Add a small random, to avoid numerical collapse of "factanal" in R
        choices += numpy.random.rand(*choices.shape) * 1e-3
        choices = list(choices)
```
The first line will take care of normalizing different, yet equivalent, votes. For example, someone voting 12344444 likely has the same intention as someone voting 12388888; likely some will even be aware that it does not matter for the voting method. So these should be treated equally. With the standard tie-averaging approach, they will be transformed to 12366666.
Now because this removes one factor (value range) from the data, it introduces a correlation into the data that kills R factanal. To make the analysis still work, I add random noise of a magnitude substantially less than the rank differences, just enough to ensure the matrix is invertible.
I don't know which approach is better (I'd argue that your approach is more sensitive to measuring whether there are people with extreme opinions that vote, e.g., 18888888); but in lieu of a formal argument for one over the other, I'd argue that observations not supported by both more likely are artifacts than true observations. Some observations still persist (AG and DH are pairs); others look quite a bit different (no F+B+fd factor; E+G decompose into two factors). I'd argue that in this analysis, the factors mostly decompose into whether people liked (1) DH best, (2) E best, (3) G best, (4) A best. This is a very plausible outcome; given that F+B were overall popular it is plausible that they are also often ranked 2nd and 3rd.

```
===== Exploratory factorial analysis

Call:
factanal(x = ~F + B + A + D + H + E + G + fd, factors = 4, data = dat,
   scores = "regression", rotation = "promax")

Uniquenesses:
    F     B     A     D     H     E     G    fd 
0.140 0.300 0.005 0.290 0.158 0.005 0.005 0.629 

Loadings:
   Factor1 Factor2 Factor3 Factor4
F  -0.501  -0.305  -0.207  -0.214 
B  -0.247  -0.568  -0.197         
A  -0.194                   1.041 
D   0.980          -0.193  -0.108 
H   1.016                  -0.154 
E  -0.154   1.133  -0.157         
G  -0.166  -0.124   1.143         
fd -0.432   0.154  -0.115  -0.321 

               Factor1 Factor2 Factor3 Factor4
SS loadings      2.580   1.752   1.478   1.285
Proportion Var   0.323   0.219   0.185   0.161
Cumulative Var   0.323   0.542   0.726   0.887

Factor Correlations:
        Factor1 Factor2 Factor3 Factor4
Factor1   1.000   0.186   0.548  -0.480
Factor2   0.186   1.000   0.203  -0.285
Factor3   0.548   0.203   1.000  -0.523
Factor4  -0.480  -0.285  -0.523   1.000

Test of the hypothesis that 4 factors are sufficient.
The chi square statistic is 6331.72 on 2 degrees of freedom.
The p-value is 0 

===== Correlation between factors #1 and #2

	Pearson's product-moment correlation

data:  sc[, 1] and sc[, 2]
t = -5.3537, df = 423, p-value = 1.417e-07
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.3389097 -0.1606376
sample estimates:
       cor 
-0.2519095 
```

But again - I'd not put too much trust on any observation not supported by both variants. Changing just a little bit to the preprocessing does produce quite different results.

**Comment by rlaboiss** [comment]
Thanks for this extended comment. As regards your proposal for changing the ballot pre-processing:

> The first line will take care of normalizing different, yet equivalent, votes. For example, someone voting 12344444 likely has the same intention as someone voting 12388888; likely some will even be aware that it does not matter for the voting method. So these should be treated equally. With the standard tie-averaging approach, they will be transformed to 12366666.

I deliberately did the opposite assumption in my study, by considering that the intention of a voter who cast the ballot =12344444= is different form the one who cast =12388888=.

As regards your final comment:

> But again - I'd not put too much trust on any observation not supported by both variants. Changing just a little bit to the preprocessing does produce quite different results.

I am a little bit surprised that slight changes in the pre-processing would cause such dramatic changes in the outcome. I would tend to say that the changes that you propose are not as slight as we think.  At any rate, robustness of the results is an important issue, and I thank you for raising it.  I will investigate the issue carefully, as time permits.

**Comment by kno10** [comment]
> I deliberately did the opposite assumption in my study, by considering that the intention of a voter who cast the ballot =12344444= is different form the one who cast =12388888=.

Yes, I figured that you intentionally treated them different; and that is why I wanted to run it with different parameters. Because when I voted, I don't make a difference here (because I know it does not affect the outcome). I doubt that many voters will choose the numbers to send a message; i.e. I do not think a voter will be asking himself "should I vote 12344444 to say they are all acceptable, or 12388888 to emphasize that I hate those alternatives". At best, they may be doing this involuntary; but then your approach overemphasize those that have extreme opinions. There is a number of votes that is just 1-------. I don't think these votes are to be treated the same as 12222222 (as they are with your preprocessing right now). Maybe 18888888 is even closer to the intention of these voters, and 15555555 is likely the best compromise to analyze them.

But that is why the effect of this tie breaking on the result is so large. In your approach, a voter that voted 1------- has much less influence compared to a user that voted 18888888 (despite being a very extreme vote). But they do actually have the same effect on the vote outcome.

If you want to capture intentions more than effect on the outcome, I'd rather analyze with respect to the placement of "further discussion". For example this vote 18888887 very clearly states that 1 is the only acceptable answer. If you want to emphasize such "extreme" votes (although the voter may just not have bothered to differentiate between the unacceptable options), then I'd rather split into values above, tied with "fd" and below, and then scale this to unit stddev, for example. But this becomes even more messy, and more detached from the voting procedure. But again, allows for quite different conclusions:
```
        choices = scipy.stats.rankdata(choices) # tie handling
        choices -= choices[-1] # shift make "fd" exactly zero.
        choices /= numpy.std(choices, ddof=0) # every voter has same stddev.
        # Add a small random, to avoid numerical collapse of "factanal" in R
        choices += numpy.random.rand(*choices.shape) * 1e-3
        choices = list(choices)
```
Now the result is
```
Uniquenesses:
    F     B     A     D     H     E     G    fd 
0.005 0.123 0.232 0.153 0.005 0.168 0.371 0.966 

Loadings:
   Factor1 Factor2 Factor3 Factor4
F           1.009                 
B  -0.109   0.324   0.329   0.574 
A   0.456           0.139   0.532 
D   0.620           0.514         
H   0.660           0.661  -0.166 
E   0.906          -0.111   0.100 
G   0.650           0.219   0.101 
fd                 -0.166         

               Factor1 Factor2 Factor3 Factor4
SS loadings      2.285   1.130   0.919   0.666
Proportion Var   0.286   0.141   0.115   0.083
Cumulative Var   0.286   0.427   0.542   0.625

Factor Correlations:
        Factor1 Factor2 Factor3 Factor4
Factor1   1.000   0.167   0.318   0.300
Factor2   0.167   1.000  -0.442  -0.194
Factor3   0.318  -0.442   1.000   0.442
Factor4   0.300  -0.194   0.442   1.000

Test of the hypothesis that 4 factors are sufficient.
The chi square statistic is 1.99 on 2 degrees of freedom.
The p-value is 0.371 
```

**Comment by rlaboiss** [comment]
Thanks for this discussion, which is indeed very important.  It essentially tells us that the interpretation of the ballots is a pretty crucial issue.  I will let the present discussion open forever (i.e. I will not close the current issue), but I am not willing to change the text of my analysis to introduce new ways of doing the ballot pre-processing.  Indeed, I clearly assumed that the choice of the numbers reflects the intention of the voters. Now, this hypothesis may be wrong, either partially or totally. However it has been clearly stated and the results should be understood as being based on this assumption.
